src/pkg/boltz.ts

```diff
@@ -0,0 +1,480 @@
+    return reverseSwap;
+  }
+
+  // check that everything is correct with data received from Boltz:
+  // - invoice
+  // - lockup address
+  // - redeem script
+  private isValidReverseSubmarineSwap({
+    invoice,
+    lockupAddress,
+    preimage,
+    claimPublicKey,
+    redeemScript,
+  }: ReverseSwap): boolean {
+    return (
+      this.correctPaymentHashInInvoice(invoice, preimage) &&
+      this.reverseSwapAddressDerivesFromScript(lockupAddress, redeemScript) &&
+      this.validateReverseSwapReedemScript(preimage, claimPublicKey, redeemScript)
+    );
+  }
+
+  // validates if invoice has correct payment hashtag
+  private correctPaymentHashInInvoice(invoice: string, preimage: Buffer): boolean {
+    const paymentHash = this.getInvoiceTag(invoice, 'payment_hash');
+    const preimageHash = crypto.sha256(preimage).toString('hex');
+    return paymentHash === preimageHash;
+  }
+
+  // return data for given tag in given invoice
+  private getInvoiceTag(invoice: string, tag: string): TagData {
+    const decodedInvoice = bolt11.decode(invoice);
+    for (const { tagName, data } of decodedInvoice.tags) {
+      if (tagName === tag) return data;
+    }
+    return '';
+  }
+
+  // validates if reverse swap address derives from redeem script
+  private reverseSwapAddressDerivesFromScript(
+    lockupAddress: string,
+    redeemScript: string
+  ): boolean {
+    const addressScript = address.toOutputScript(lockupAddress);
+    const addressScriptASM = script.toASM(script.decompile(addressScript) || []);
+    const sha256 = crypto.sha256(Buffer.from(redeemScript, 'hex')).toString('hex');
+    const expectedAddressScriptASM = `OP_0 ${sha256}`; // P2SH
+    return addressScriptASM === expectedAddressScriptASM;
+  }
+
+  // validates if we can redeem with this redeem script
+  private validateReverseSwapReedemScript(
+    preimage: Buffer,
+    pubKey: string,
+    redeemScript: string
+  ): boolean {
+    const scriptAssembly = script
+      .toASM(script.decompile(Buffer.from(redeemScript, 'hex')) || [])
+      .split(' ');
+    const cltv = scriptAssembly[10];
+    const refundPubKey = scriptAssembly[13];
+    const expectedScript = [
+      'OP_SIZE',
+      '20',
+      'OP_EQUAL',
+      'OP_IF',
+      'OP_HASH160',
+      crypto.hash160(preimage).toString('hex'),
+      'OP_EQUALVERIFY',
+      pubKey,
+      'OP_ELSE',
+      'OP_DROP',
+      cltv,
+      'OP_NOP2',
+      'OP_DROP',
+      refundPubKey,
+      'OP_ENDIF',
+      'OP_CHECKSIG',
+    ];
+    return scriptAssembly.join() === expectedScript.join();
+  }
+
+  // validates redeem script is in expected template
+  private validateSwapReedemScript(redeemScript: string, refundPublicKey: string) {
+    const scriptAssembly = script
+      .toASM(script.decompile(Buffer.from(redeemScript, 'hex')) || [])
+      .split(' ');
+    const boltzHash = scriptAssembly[4];
+    const cltv = scriptAssembly[6];
+    const preimageHash = scriptAssembly[1];
```

Comment:
Get this preimage hash not from our script but the invoice the user provided

Reply:
It's ok, the preimageHash is verified before:
https://github.com/vulpemventures/marina/pull/482/files#diff-354258afee577a66599ab8256034996a1b1ea9cfb29d046d38b444a55b4402deR316